client/game_logic.py

```diff
@@ -0,0 +1,190 @@
+import threading
+import time
+import math
+import random
+from itertools import accumulate
+
+MAX_SCORE = 100
+
+
+class Game(threading.Thread):
+    def __init__(self, pressed_keys: list, pressed_keys_lock: threading.Lock):
+        super().__init__()
+        self.pressed_key = pressed_keys
+        self.pressed_keys_lock = pressed_keys_lock
+        self.package_number = 0
+        self.score = 0
+        self.combo = 1
+        self.number_of_moles = 9
+        self.values_changed = False
+        self.mole_life_time = [2, 3]
+        self.max_moles_amount = 3
+        self.last_print_time = 0
+        self.moles_status = [[0, [0, 0, 0]]
+                             for _ in range(self.number_of_moles)]  # [status, alive_time]
+        self.moles_lock = threading.Lock()
+        self.end = False
+
+    def run(self):
+        self.last_print_time = time.time()
+        try:
+            while not self.end:
+                self.values_changed = False
+                self.whack_moles()
+                self.generate_new_moles()
+                self.moles_progression()
+                if self.values_changed and self.last_print_time + 1 < time.time():
+                    self.print_board()
+                    self.package_number += 1
+                    self.last_print_time = time.time()
+                    if self.score >= MAX_SCORE:
+                        self.end = True
+                    print(self.create_bits_package())
+                time.sleep(0.2)
+        except KeyboardInterrupt:
+            return
+
+    def generate_mole_lifetime(self) -> list:
+        value = random.uniform(self.mole_life_time[0], self.mole_life_time[1])
+        return [time.time() + value * (x + 1) / 3 for x in range(3)]
+
+    def set_new_mole(self, pos: int):
+        self.moles_status[pos] = [3, self.generate_mole_lifetime()]
+
+    def hide_mole(self, pos: int):
+        self.moles_status[pos] = [0, 0]
+
+    def moles_progression(self):
+        self.moles_lock.acquire()
+        op_start_time = time.time()
+        for idx, mole in enumerate(self.moles_status):
+            for state, checked_time in enumerate(mole[1][::-1]):
+                if checked_time < op_start_time:
+                    self.moles_status[idx][0] = state
+                    self.values_changed = True
+                    break
```

Comment:
Shouldn't alive_time be altered/erased when moles state has changed?

Reply:
Nope, thats why it has 3 time gates. Each one means 1/3 of life time. Thats why i check it in reversed order